### PR TITLE
Fix issue where ops are sent out of order on connect

### PR DESCRIPTION
**Pull request by kian-thompson**
There are observed cases where ops are sent out of order upon connecting. To fix this, we need to flush the Outbox when connecting so the ops are replayed in the original order.

The sequence of events is the following:
1. ContainerRuntime is disconnected (whether from the start or not doesn't matter)
2. ops are submitted to Outbox
3. batch is flushed and ops are stored in PendingStateManager
4. ops are not sent over wire since we are disconnected
5. more ops are submitted to Outbox
6. we receive join op for our client
7. we replay pending ops before the current batch is flushed
8. when resubmitting an op, we see that the Outbox's mainBatch has a lower ref sqn than the DeltaManager (last op was the "join")

[AB#4004](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4004)

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -2039,6 +2037,12 @@ export class ContainerRuntime
 		// There might be no change of state due to Container calling this API after loading runtime.
 		const changeOfState = this._connected !== connected;
 		const reconnection = changeOfState && !connected;
+
+		// We need to flush the ops currently collected by Outbox to the PendingStateManager to preserve original order upon calling "replayPendingStates"
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2041]: Should we do this closer to the `replayPendingStates` below? I mean in the same conditional?

Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2041]: Oh, I think not, as by the time that happens, connected is true, so the outbox will actually send them. Can you add a comment to be more explicit here and call out the fact that this code relies on the shouldSend parameter provided to the outbox (connected = false) and the idea is to roundtrip all batches to the pendingstatemanager via the outbox without actually sending it?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -1606,8 +1606,6 @@ export class ContainerRuntime
 				!readonly || !this.connected,
 				0x125 /* "Unsafe to transition to read-only state!" */,
 			);
-
-			this.replayPendingStates();
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1610]: Is the comment above still relevant? cc @vladsud

**Comment by anthony-murphy** [comment]
could this be related to #16600

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (kian-thompson) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1610]: As of "2.0.0-internal.1.2.0", we set the connection state in Container on the DeltaManager's "readonly" event.
https://github.com/microsoft/FluidFramework/pull/11655

Maybe this call should be kept for compat with lts version? This place would also need similar Outbox flushing.

Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1610]: Why is this removed? "readonly" doesn't seem to have any related to this bug and the scenario in the PR description.

**Review by andre4i — APPROVED** [review]
